### PR TITLE
Run Length Encoding (RLE) support

### PR DESCRIPTION
**Pull request by AaronM04**
I wasn't planning on submitting this now but I might as well because I'm working more slowly on the remaining 5% of this.

Remaining parts to be in a separate PR:
* some more unit tests for my later stuff (in particular `Universe::apply` has zero unit tests 😬 )
* a consistent error type
* better documentation comments

The other part of this is integration into the game client in the other repository ( https://github.com/conwayste/conwayste/tree/aaron/rle-libconway ), which I'm going to start on while you review this. BTW, this is six months of work so feel free to take your time on this, and feel free to ask questions!

### CODE REVIEW

File: src/bits.rs [hunk 1]
```diff
@@ -0,0 +1,723 @@
...
+    #[inline]
+    pub fn modify_bits_in_word(&mut self, row: usize, word_col: usize, mask: u64, op: BitOperation) {
+        match op {
+            BitOperation::Set    => self[row][word_col] |=  mask,
+            BitOperation::Clear  => self[row][word_col] &= !mask,
+            BitOperation::Toggle => self[row][word_col] ^=  mask,
+        }
+    }
+
+    // Sets or clears a rectangle of bits. Panics if Region is out of range.
```
Comment (Manghi) [inline src/bits.rs]: Needs an additional forward slash for docs. No need to fix here unless you want to, since you want to do additional documentation updates in the next PR.

File: src/bits.rs [hunk 2]
```diff
@@ -0,0 +1,723 @@
...
+        }
+    }
+
+    /// Get a Region of the same size as the BitGrid.
+    pub fn region(&self) -> Region {
+        Region::new(0, 0, self.width(), self.height())
+    }
+
+    /// Clear this BitGrid.
+    pub fn clear(&mut self) {
```
Comment (Manghi) [inline src/bits.rs]: Why not iterate over all the components and just set to zero? 
Pseudo: `self[i][j] = 0 for i,j to width,height`

File: src/bits.rs [hunk 3]
```diff
@@ -0,0 +1,723 @@
...
+        let region = {
+            self.region()
+        };
+        self.modify_region(region, BitOperation::Clear);
+    }
+}
+
+
+impl Index<usize> for BitGrid {
```
Comment (Manghi) [inline src/bits.rs]: :smile:

File: src/bits.rs [hunk 4]
```diff
@@ -0,0 +1,723 @@
...
+        self.0.len()
+    }
+
+    /// _visibility is ignored, since BitGrids have no concept of a player.
+    fn write_at_position(&mut self, col: usize, row: usize, ch: char, _visibility: Option<usize>) {
+        let word_col = col/64;
+        let shift = 63 - (col & (64 - 1));
+        match ch {
+            'b' => {
+                self[row][word_col] &= !(1 << shift)
```
Comment (Manghi) [inline src/bits.rs]: use `self.modify_bits_in_word(...)` for here and line 330?

File: src/rle.rs [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+                return Err(format!("unexpected term in header line: {:?}", term));
+            }
+            map.insert(parts[0], parts[1]);
+        }
+        if !map.contains_key("x") || !map.contains_key("y") {
+            return Err(format!("header line missing `x` and/or `y`: {:?}", line));
+        }
+        let x = usize::from_str(map.get("x").unwrap()).map_err(|e| format!("Error while parsing x: {}", e))?;
+        let y = usize::from_str(map.get("y").unwrap()).map_err(|e| format!("Error while parsing y: {}", e))?;
+        let rule =  map.get("rule").map(|s: &&str| (*s).to_owned());
```
Comment (Manghi) [inline src/rle.rs:139]: This will return `None` if the rule portion is omitted. Should we error or is that expected?
```
x=36, y=9
24bo$22bobo$12b2o6b2o12b2o$11bo3bo4b2o12b2o$2o8bo5bo3b2o$2o8bo3bob2o4bobo$10bo5bo7bo$11bo3bo$12b2o!
```

**Comment by Manghi** [comment]
Other than commented above, `rle.rs` and `bits.rs` look great. Wonderful job! I still need to look at the `universe.rs` changes.

*Minor nitpick: Since it is a new file, what do you think about naming `bits.rs` to `grids.rs`?*

File: src/rle.rs (hunk 5 above)
Comment (AaronM04) [reply src/rle.rs:139]: That's expected.

File: src/bits.rs (hunk 4 above)
Comment (AaronM04) [reply src/bits.rs]: you got it! :grin:

File: src/bits.rs (hunk 2 above)
Comment (AaronM04) [reply src/bits.rs]: Cool, sounds good, thanks for the suggestion.

Added bonus: I realized I was missing a unit test for this so I added that.

File: src/bits.rs (hunk 1 above)
Comment (AaronM04) [reply src/bits.rs]: Fixed!
